Handle conditionalDependencies

andreypopp opened the pull request:
I propose to add `conditionalDependencies` (`optionalDependencies` name is
already taken by npm and doesn't have the needed mechanics behind it).

The suggestion is to use the same semantics as for `peerDependencies` but not
fail if those packages mentioned in conditionalDependencies are absent from
installation.

If package `optdep` is listed in `conditionalDependencies` for the package `pkg`:

* `optdep` is not being installed automatically (same as with
  `peerDependencies`) for the `pkg`
* Unlike with peerDependencies it is not an error to have no `optdep` installed
  at all.
* If `optdep` is present in the installation (required by some of packages
  above) then it's being pulled into `pkg`'s env.

Fixes #46

andreypopp committed:
Handle conditionalDependencies

I propose to add `conditionalDependencies` (`optionalDependencies` name is
already taken by npm and doesn't have the needed mechanics behind it).

The suggestion is to use the same semantics as for `peerDependencies` but not
fail if those packages mentioned in conditionalDependencies are absent from
installation.

If package `optdep` is listed in `conditionalDependencies` for the package `pkg`:

* `optdep` is not being installed automatically (same as with
  `peerDependencies`) for the `pkg`
* Unlike with peerDependencies it is not an error to have no `optdep` installed
  at all.
* If `optdep` is present in the installation (required by some of packages
  above) then it's being pulled into `pkg`'s env.

Fixes #46

andreypopp commented on src/installCommand/index.js:
```diff
@@ -108,9 +109,46 @@ const installationSpec = {
+      if (!installedPackages) {
+        return;
+      }
+      await linkConditionalDependencies(installedPackages);
```
This is needed because of how strict pnpm is — it only links those packages which are mentioned in deps. This is why we needed to modify algo to include `conditionalDependencies`.

jordwalke replied:
Should this be done via some kind of `installWillLink` API hook? Is it fragile or discouraged to "link" in this way?

jordwalke replied:
I could imagine an API like:

```
installWillLink  : tentativeListOfLinking => augmentedListOfLinking
```

Where you get the "tentative list" passed as an arg, and then you can return an augmented/reduced version of that list to help influence exactly what is linked.

andreypopp replied:
Yeah, that's the idea eventually.

andreypopp replied:
Right now I'm not sure what other hooks we need from pnpm and in what form — I'm also want to dog food pnpm plugin API a bit before submitting a merge proposal to pnpm.

andreypopp commented:
I propose pulling this in but not documenting, at least for now. It allows much more lightweight builds for common packages which have different backends — like `cohttp` for example. While we can have it now (if we merge this) we can decide on how we might want to see this as a public API.

jordwalke commented:
Yes, I like your plan @andreypopp - please merge when ready.

andreypopp commented:
Hm... actually found a downside of this approach — in Esy's current form it requires an app which depend on lwt to depend on `base-unix` and `base-thread`. This is what `lwt` gets converted to:
```
      ...
      "esy": {
        "build": [
          "make setup", 
          "ocaml setup.ml -configure ... --${base_unix_enable:-disable}-unix --${base_threads_enable:-disable}-preemptive", 
          "make build", 
          "make install", 
          "(opam-installer --prefix=$cur__install || true)"
        ], 
      }
      ...
```
Not sure how to model that in Esy...

andreypopp commented:
We could probably use `ocamlfind` to query if `unix`/`threads` are installed a set env accordingly. That way we don't need `base-*` packages at all in dep graph.

andreypopp committed:
install: support "files" inside directories

That allows to put files within directories within opam packages:

    "opam": {
      "files": {
        "name: "_esy/file",
        "content": "..."
      }
    }

Note the `_esy/file`.

andreypopp committed:
build-eject: do not expand build command

By quoting with "'" we don't allow build command to expand early. That means
some earlier commands can affect environment for later commands. I think this is
expected as that's how dependency-env worked (remember nopam?).

andreypopp committed:
sandbox: fix error messages

andreypopp committed:
opam-conversion: model unix/threads directly in env

Opam models unix/treads as packages we can't do the same as we can eject build
into another platform. Instead we query ocamlfind for unix/threads and set opam
variables accordingly.

andreypopp committed:
opam-conversion: rm base-{unix,threads} from list of converted packages

andreypopp commented:
ready for review, @jordwalke 

In this PR we model `unix`/`threads` deps dynamically by querying `ocamlfind` for those libs.

andreypopp committed:
opam-conversion: set base_{unix,threads}_installed

This was missing in previous commit.
